### PR TITLE
[MIRROR] Replace hardcoded z-level numbers with a trait system

### PR DESCRIPTION
**Pull request by ghost**
Original PR: https://github.com/tgstation/tgstation/pull/34090
--------------------
:cl:
refactor: Information about the properties of z-levels is now less hardcoded than before.
/:cl:

Done in this PR:

* Add trait system from #32037 and prepare file layout for the future
* Update defines introduced in #33829 to use traits
* Define traits for the maps in `basemap.dm`, the autogenerated deep space, and away missions
* Use traits to handle bombcap increase on Lavaland and decrease on Reebe
* Use traits to know which z-levels to spawn lava or space ruins on
* Replace `GLOB.station_z_levels` entirely by traits
* Eliminate some but not all hardcoded z-level checks
* Add a "Debug Z-Levels" verb which shows all z-level info and a map of deep space

I specifically tested space transitions to make sure they still work. Generally speaking, no player-facing behavior should have changed. There's a lot more to go on the z-level stuff but this is another midpoint where PR-worthy work has been done and everything is in a functional state.

### CODE REVIEW

**Comment by Pyko1** [comment]
added/removed files to hippiestation.dme

**Comment by Pyko1** [comment]
`hippiestation/code/game/objects/items/implants/implant_teleporter.dm:19:error: ZLEVEL_CENTCOM: undefined var
hippiestation/code/game/objects/items/implants/implant_teleporter.dm:47:error: ZLEVEL_CENTCOM: undefined var
code/modules/mining/equipment/survival_pod.dm:68:error: ZLEVEL_MINING: undefined var
hippiestation/code/game/gamemodes/shadowling/shadowling.dm:263:error: GLOB.station_z_levels: undefined var
hippiestation/code/game/machinery/telecomms/machines/receiver.dm:27:error: GLOB.station_z_levels: undefined var
hippiestation.mdme.dmb - 5 errors, 0 warnings (1/12/18 9:16 am)
mv: cannot stat ‘hippiestation.mdme.dmb’: No such file or directory
`

**Comment by Pyko1** [comment]
Someone better check my changes in case i misunderstood how shit works.

tl;dr

replaced 
`if(imp_in.z != ZLEVEL_CENTCOM)`
with
`if(!is_centcom_level(imp_in.z))`

`(T.z in GLOB.station_z_levels))`
replaced with
`is_station_level(T.z)`

`(O.z in GLOB.station_z_levels)`
Replaced with
`is_station_level(O.z)`

`(T.z != ZLEVEL_MINING && T.z != ZLEVEL_LAVALAND)`
replaced with
`is_mining_level(T.z)`

File: code/modules/mining/equipment/survival_pod.dm [hunk 1]
```diff
@@ -65,7 +65,7 @@
 		playsound(get_turf(src), 'sound/effects/phasein.ogg', 100, 1)
 
 		var/turf/T = deploy_location
-		if(T.z != ZLEVEL_MINING && T.z != ZLEVEL_LAVALAND)//only report capsules away from the mining/lavaland level
+		if(!is_mining_level(T.z))//only report capsules away from the mining/lavaland level
```
Comment (JohnGinnane) [inline code/modules/mining/equipment/survival_pod.dm]: Shouldn't we also check for lavaland?

Comment (Pyko1) [reply code/modules/mining/equipment/survival_pod.dm]: DECLARE_LEVEL("Lavaland", UNAFFECTED, list(ZTRAIT_MINING = TRUE, ZTRAIT_LAVA_RUINS = TRUE, ZTRAIT_BOMBCAP_MULTIPLIER = 3)),\

Comment (Pyko1) [reply code/modules/mining/equipment/survival_pod.dm]: as far as I understand lavaland should check as mining level

Comment (Pyko1) [reply code/modules/mining/equipment/survival_pod.dm]: Even tho main question is why this isn't in hippie modules

Comment (Pyko1) [reply code/modules/mining/equipment/survival_pod.dm]: oh for some reason tg changes didn't just mirror properly to this file.

Ill overwrite with up to date tg.

**Comment by Pyko1** [comment]
looks like our code/modules/mining/equipment/survival_pod.dm
had some undocumented changes that should had been prolly in hippie folder.

**Comment by Pyko1** [comment]
updated survival pod to up to date tg and added our chances back with comments

**Review by Carbonhell — APPROVED** [review]
